### PR TITLE
Fix empty short string encoding

### LINKED ISSUE #4087: `toStringWithFallback` misbehaves for an empty string
I'm sorry for the informal format of this issue, but it's a really small one I've randomly found while reading code.

`toStringWithFallback` checks if the short string encoding failed by running `if (length(value) > 0) {`, but this doesn't take into account a genuinely empty string, which encodes in `toShortStringWithFallback` just fine, without a fallback: `if (bytes(value).length < 32) {` ` return toShortString(value);`. The result is that if a genuinely empty ShortString is used with a non-empty fallback storage string, the storage will be read. This wastes gas, but may also return an invalid value if storage isn't properly cleaned.

To solve this IMO we should encode differently a fallen-back ShortString in `toShortStringWithFallback`, e.g. by setting its length to 255.

### PR DESCRIPTION
**Pull request by Amxx**
Fixes #4087

#### PR Checklist

- [x] Tests

### CODE REVIEW

File: contracts/utils/ShortStrings.sol [hunk 1]
```diff
@@ -44,7 +45,7 @@ library ShortStrings {
         if (bstr.length > 31) {
             revert StringTooLong(str);
         }
-        return ShortString.wrap(bytes32(uint256(bytes32(bstr)) | bstr.length));
+        return ShortString.wrap(bytes32(uint256(bytes32(bstr)) | (bstr.length + 1)));
```
[the remark this answers is not in the document]
Comment (Amxx) [reply contracts/utils/ShortStrings.sol]: done

**Comment by frangio** [comment]
This presents a question. What should be the value of an uninitialized `ShortString`? This PR is currently making that an invalid state (attempt to decode results in `InvalidShortString`), and I think intuitively people would expect it to be an empty string like the native `string` type.

I think a ShortString should only become invalid should be when you use `setWithFallback` and the function has to fall back to storage. `setWithFallback` should set a sentinel value (e.g. length = 255) to indicate that the value must be read from storage with `getWithFallback`.

File: contracts/utils/ShortStrings.sol [hunk 2]
```diff
@@ -32,7 +32,10 @@ type ShortString is bytes32;
  * ```
  */
 library ShortStrings {
+    bytes32 constant _INVALID_SHORT_STRING = 0x00000000000000000000000000000000000000000000000000000000000000FF;
```
Comment (frangio) [inline contracts/utils/ShortStrings.sol:35]:
```suggestion
    bytes32 private constant _INVALID_SHORT_STRING = 0x00000000000000000000000000000000000000000000000000000000000000FF;
```

Comment (frangio) [inline contracts/utils/ShortStrings.sol:35]: I would call this something like `_FALLBACK_SENTINEL`.

**Comment by CodeSandwich** [comment]
Thank you for fixing it! From my side it looks good, but obviously I'm not a maintainer.

**Review by frangio — APPROVED** [review]

**Comment by frangio** [comment]
Looks good to me.

The new logic makes it more evident that `lengthWithFallback` might make sense. But I wouldn't add that in this PR.

**Pull request merged**